File: codeforces/1800~1899/1870/d.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

void solve() {
  int n = 0;
  cin >> n;
  vector<int> c(n);
  for (int i = 0; i < n; ++i) cin >> c[i];
  int k = 0;
  cin >> k;
  for (int i = n-2; i >= 0; --i) {
    c[i] = min(c[i], c[i+1]);
  }
  
  vector<int> a(n);
  a[0] = k / c[0];
  int x = k - k / c[0] * c[0];
  for (int i = 1; i < n; ++i) {
    if (c[i] == c[i-1]) {
      a[i] = a[i-1];
    } else {
      int require = c[i] - c[i-1];
      a[i] = min(a[i-1], x / require);
      x = max(0, x - require * a[i]);
    }
  }

  for (int i = 0; i < n; ++i) {
    cout << a[i] << " ";
  }
  cout << endl;
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  int T = 0;
  cin >> T;
  for (int t = 0; t < T; ++t) {
    solve();
  }
  return 0;
}

File: codeforces/1800~1899/1870/d.rs
use std::io::{stdin, stdout, BufWriter, Write};
use std::str::FromStr;
use std::fmt::Debug;
use std::cmp::{min, max};

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn solve() {
    let stdout = stdout();
    let mut writer = BufWriter::new(stdout.lock());

    let n = read_int::<usize>();
    let mut c = read_vec::<i32>();
    let k = read_int::<i32>();

    for i in (0..n-1).rev() {
        c[i] = min(c[i], c[i+1]);
    }

    let mut a = vec![0; n];
    a[0] = k / c[0];
    let mut x = k - k / c[0] * c[0];
    for i in 1..n {
        if c[i] == c[i-1] {
            a[i] = a[i-1];
        } else {
            let add = c[i] - c[i-1];
            a[i] = min(a[i-1], x / add);
            x = max(0, x - add * a[i]);
        }
    }

    for i in 0..n {
        write!(writer, "{} ", a[i]).ok();
    }
    writeln!(writer).ok();
}

fn main() {
    let n = read_int::<i32>();
    for _t in 0..n {
        solve();
    }
}
